File: code/ch19/19.1.5. 189.rotate.py
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        # 首尾交换法
        def reverse(list: List[int], start: int, end: int) -> None:
            while start < end:
                t = list[start]
                list[start] = list[end]
                list[end] = t
                start += 1
                end -= 1

        n = len(nums)
        offset = k % n
        if offset == 0:
            return
        reverse(nums, 0, n - offset - 1)
        reverse(nums, n - offset, n - 1)
        reverse(nums, 0, n - 1)

File: code/ch19/19.1.5. 189.rotate.cpp
#include <vector>
using namespace std;

class Solution {
public:
  void reverse(vector<int>& arr, int start, int end) {
    while (start < end) {
      int t = arr[start];
      arr[start] = arr[end];
      arr[end] = t;
      start += 1;
      end -= 1;
    }
  }
  void rotate(vector<int>& nums, int k) {
    int n = nums.size(), offset = k % n;
    if (offset == 0) return;
    reverse(nums, 0, n - offset - 1);
    reverse(nums, n - offset, n - 1);
    reverse(nums, 0, n - 1);
  }
};
